### PR TITLE
Port //daml-lf/(parser|validation) to Scala 2.13

### PR DESCRIPTION
**Pull request by cocreature**
changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/parser/src/main/scala/com/digitalasset/daml/lf/testing/parser/ExprParser.scala [hunk 1]
```diff
@@ -44,7 +44,7 @@ private[parser] class ExprParser[P](parserParameters: ParserParameters[P]) {
       (id ^? builtinFunctions) ^^ EBuiltin |
       caseOf |
       id ^^ EVar |
-      `(` ~> expr <~ `)`
+      (`(` ~> expr <~ `)`)
```
Comment (cocreature) [inline daml-lf/parser/src/main/scala/com/digitalasset/daml/lf/testing/parser/ExprParser.scala:47]: This is to work around https://github.com/scala/bug/issues/12071. Afaict, the warning is a false positive but adding a few parentheses seems easy enough.

File: daml-lf/parser/src/main/scala/com/digitalasset/daml/lf/testing/parser/Lexer.scala [hunk 2]
```diff
@@ -67,7 +67,7 @@ private[parser] object Lexer extends RegexParsers {
       "=" ^^^ `=` |
       "_" ^^^ Token.`_` |
       "|" ^^^ `|` |
-      """[a-zA-Z\$_][\w\$]*""".r ^^ (s => keywords.getOrElse(s, Id(s))) |
+      """[a-zA-Z_\$][\w\$]*""".r ^^ (s => keywords.getOrElse(s, Id(s))) |
```
Comment (cocreature) [inline daml-lf/parser/src/main/scala/com/digitalasset/daml/lf/testing/parser/Lexer.scala:70]: `$_` now triggers the missing interpolator warning

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Collision.scala [hunk 3]
```diff
@@ -35,8 +35,8 @@ private[validation] object Collision {
     } throw ECollision(pkgId, entity1, entity2)
 
   private def namedEntitiesFromPkg(
-      modules: Traversable[(ModuleName, Ast.Module)]
```
Comment (cocreature) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Collision.scala:38]: Traversable is dead in 2.13

File: daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/SpecUtil.scala [hunk 4]
```diff
@@ -26,7 +26,7 @@ private[validation] object SpecUtil {
     '∀' -> "forall",
     '→' -> "->",
     '←' -> "<-",
-    '₁' → "_1",
```
Comment (cocreature) [inline daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/SpecUtil.scala:29]: deprecated and using unicode in one case doesn’t seem helpful.

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/TypeSubst.scala [hunk 5]
```diff
@@ -32,7 +33,7 @@ private[validation] object TypeSubst {
     }
 
   private def freshTypeVarName(fv: Set[TypeVarName]): TypeVarName =
-    Stream
+    LazyList
```
Comment (remyhaemmerle-da) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/TypeSubst.scala:36]: :+1:

File: daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/SpecUtil.scala (hunk 4 above)
Comment (remyhaemmerle-da) [reply daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/SpecUtil.scala:29]: :+1:

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/ExprTraversable.scala [hunk 6]
```diff
@@ -158,36 +140,31 @@ private[validation] object ExprTraversable {
...
-    new Traversable[Expr] {
-      def foreach[U](f: Expr => U): Unit = {
-        module.definitions.values.foreach(that.foreach(_, f))
-        module.templates.values.foreach(that.foreach(_, f))
-      }
+  def apply(module: Module): Iterable[Expr] =
+    new Iterable[Expr] {
+      override def iterator: Iterator[Expr] =
+        module.definitions.values.iterator.flatMap(ExprTraversable.iterator(_)) ++
+          module.templates.values.iterator.flatMap(ExprTraversable.iterator(_))
```
Comment (remyhaemmerle-da) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/ExprTraversable.scala:168]: So more object allocation. :(.
It should be fine for the uses cases, though.

**Review by remyhaemmerle-da — APPROVED** [review]
Looks good to me.

Nonetheless, I would advocate to rename `com.daml.lf.validation.traversable` to `com.daml.lf.validation.iterable` and substitute "traversable" by "iterable" in all the names you can found in this package.

**Comment by cocreature** [comment]
> Nonetheless, I would advocate to rename `com.daml.lf.validation.traversable` to `com.daml.lf.validation.iterable` and substitute "traversable" by "iterable" in all the names you can found in this package.

Good point done

**Review by remyhaemmerle-da — APPROVED** [review]
:+1:

**Review by S11001001 — APPROVED** [review]

**Pull request merged**